README
= Introduction =

_Youtube-upload_ is a command-line script that uploads videos to Youtube. If a video does not comply with Youtube limitations (<15mins for a normal user) you must split it before using ffmpeg or any other tool. _Youtube-upload_ should work on any platform (GNU/Linux, BSD, OS X, Windows, ...) that runs Python.

= Dependencies =

  * [http://www.python.org python 2.6 or 2.7]
  * [http://code.google.com/p/gdata-python-client python-gdata] (>= 1.2.4)

Note: You must have logged in at least once into your Youtube account prior to uploading any videos. 

= Download & Install =

  * [http://code.google.com/p/youtube-upload/downloads/list Stable release]:

{{{
$ wget http://youtube-upload.googlecode.com/files/youtube-upload-VERSION.tgz
$ tar xvzf youtube-upload-VERSION.tgz
$ cd youtube-upload-VERSION
$ sudo python setup.py install
}}}

  * [http://code.google.com/p/youtube-upload/source/checkout From repository]:

{{{
$ svn checkout http://youtube-upload.googlecode.com/svn/trunk/ youtube-upload
$ cd youtube-upload
$ sudo python setup.py install
}}}

  * If you don't want (or you can't) install software on the computer, run it directly from sources:

{{{
$ cd youtube-upload-VERSION
$ python youtube_upload/youtube_upload.py ...
}}}

= Usage examples =

* Upload a video:

{{{
$ youtube-upload --email=myemail@gmail.com --password=mypassword \
                 --title="A.S. Mutter" --description="A.S. Mutter plays Beethoven" \
                 --category=Music --keywords="mutter, beethoven" anne_sophie_mutter.flv
www.youtube.com/watch?v=pxzZ-fYjeYs
}}}

* Upload a video with a description from file (_description.txt_):

{{{
$ youtube-upload --email=myemail@gmail.com --password=mypassword \
                 --title="A.S. Mutter" --description="$(< \description.txt)" \
                 --category=Music --keywords="mutter, beethoven" anne_sophie_mutter.flv
www.youtube.com/watch?v=pxzZ-fYjeYs
}}}

* Upload the video using the Youtube API:

{{{
$ youtube-upload --api-upload [OTHER OPTIONS] file.flv
}}}

If you set explicitly the {{{--api-upload}}} options or {{{pycurl}}} isn't installed, the original Youtube API will be used to upload the video file. This method is not recommended because it does not shows the nice progressbar.

* Upload a splited video:

{{{
$ youtube-upload [OPTIONS] --title="TITLE" video.part1.avi video.part2.avi
www.youtube.com/watch?v=pxzZ-fYjeYs # title: TITLE [1/2]
www.youtube.com/watch?v=pxzZ-fYsdff # title: TITLE [2/2]
}}}

* Add a video to a playlist:

{{{
$ youtube-upload [OPTIONS] --add-to-playlist=http://gdata.youtube.com/feeds/api/playlists/7986C428284A40A1 http://www.youtube.com/watch?v=Zpqu97l3G1U
}}}

Note that the argument must be the video URL (not a local file) and the playlist is the URL of the feed (with no prefix "PL").

* Get available categories:

{{{
$ youtube-upload --get-categories
Tech Education Animals People Travel Entertainment Howto Sports Autos Music News Games Nonprofit Comedy Film
}}}

* Split a video with _ffmpeg_

Youtube currently limits videos to <2Gb and <15' for almost all users. You can use the Bash example script to split it before uploading:

{{{
$ bash examples/split_video_for_youtube.sh video.avi
video.part1.avi
video.part2.avi
video.part3.avi
}}}

* Upload videos with _curl_

The script uses pycurl by default (when available) to upload the video, but if you need to tweak the upload parameters take a look at the Bash script included with the sources ([http://code.google.com/p/youtube-upload/source/browse/trunk/examples/upload_with_curl.sh examples/upload_with_curl.sh]). This command, for example, would upload _somevideo.flv_ with a limit rate of 100KBytes:

{{{
$ youtube-upload --get-upload-form-info [OPTIONS] | bash examples/upload_with_curl.sh --limit-rate 100k
}}}

* Upload a non-public (private/unlisted) video:

{{{
$ youtube-upload --private ...
}}}

{{{
$ youtube-upload --unlisted ...
}}}

* Get password from standard input

{{{
$ cat password.txt | youtube-upload -p - ....
}}}

* Use a HTTP proxy

Set environment variables *http_proxy* and *https_proxy*:

{{{
$ export http_proxy=http://user:password@host:port
$ export https_proxy=http://user:password@host:port
$ youtube-upload ....

}}}

* Update metadata (title or description).

Set new title for video:

{{{
$ youtube-upload --email=EMAIL --password=PASSWORD --update-metadata --title "new title" www.youtube.com/watch?v=pxzZ-fYjeYs
}}}

= Caveats =

== Validate your credentials == 
Before you upload a video using the script, do it manually to make sure everything is ok (you have a valid user/password and an associated channel). See issue57 for more details.

== Using two-step verification ==

If you are using [https://support.google.com/accounts/answer/180744?hl=en two-step verification] issue an application-specific password and use it in the script as if it was your real password. See [https://code.google.com/p/youtube-upload/issues/detail?id=111 issue111] for more details.

== Sign in using App Passwords ==

An App password is a 16-digit passcode that gives an app or device permission to access your Google Account. If you use 2-Step-Verification and are seeing a “password incorrect” error when trying to access your Google Account, an App password may solve the problem. 

https://support.google.com/accounts/answer/185833?hl=en

== Progressbar ==

You need both packages *python-progressbar* and *pycurl* installed to see the progress bar when uploading a video.

= Feedback =

Use the [http://code.google.com/p/youtube-upload/issues/ issues tracker] to report bugs or suggest improvements.


CHANGELOG
youtube-upload (0.8.0) stable; urgency=low

  * Update to APIv3 and Oauth2.

youtube-upload (0.7.4) stable; urgency=low

  * issue131: New options --nocomments, --noratings, --recorded
  * issue136: Issue with split_video_for_youtube.sh
  * add keywords to option update-metadata

 --  <pyarnau@gmail.com>  Mon, 06 May 2013 14:20:04 +0200

youtube-upload (0.7.3) stable; urgency=low

  * issue101: don't fail when no description is provided 
  * issue107: add options --hide-progressbar

 --  <pyarnau@gmail.com>  Mon, 06 May 2013 14:20:04 +0200

youtube-upload (0.7.2) stable; urgency=low

  * Change --email option help message.
  * Allow escaped chars in description
  * Add -update-metadata (basic support: title/description)

 --  <pyarnau@gmail.com>  Wed, 31 May 2012 18:58:44 +0200

youtube-upload (0.7.1) stable; urgency=low

  * Format checks on video URL and playlists.
  * Add option --unlisted (issue60).
  * Fix progressbar on finish.

 --  <pyarnau@gmail.com>  Wed, 1 May 2012 18:58:44 +0200

youtube-upload (0.7) stable; urgency=low

  * Prompt password when not specified in command line.
  * Meaningful error codes. 
  * Update to progressbar v2.3.

 --  <pyarnau@gmail.com>  Thu, 11 Aug 2011 09:54:47 +0200

youtube-upload (0.6.2) unstable; urgency=low

  * Add option to create playlist
  * Add pycurl + progression bar

 --  <pyarnau@gmail.com>  Sat, 23 Apr 2011 20:06:34 +0200

youtube-upload (0.6.1) unstable; urgency=low

  * fix issue with required options

 --  <pyarnau@gmail.com>  Fri, 25 Mar 2011 21:01:29 +0100

youtube-upload (0.6) unstable; urgency=low

  * Remove ffmpeg spliting from Python script (orthogonal task)
  * Move all arguments to options except video paths.
  * Bash script to split videos: split_video_for_youtube.sh

 --  <pyarnau@gmail.com>  Thu, 10 Mar 2011 18:50:45 +0100

youtube-upload (0.5) unstable; urgency=low

  * Update to new Youtube limits (15mins, 2Gb) 
  * Add curl options to bash script
  * Fix encoding for debug (issue10)
  * Fix location bug (issue11)
  * Fix keywords splitting (issue13)
  * Add --wait-processed option (issue19)

 --  <pyarnau@gmail.com>  Wed, 29 Sep 2010 11:31:18 +0200

youtube-upload (0.4) unstable; urgency=low

  * issue7: fix for gdata 2.0.10
  * issue9: password can be passed through <stdin>
  * Add --location option
  * Add --private option
  * Add --playlist-uri option
  * Add Bash example: Upload a video using curl

 --  <pyarnau@gmail.com>  Tue, 29 Jun 2010 19:45:40 +0200

youtube-upload (0.3) unstable; urgency=low

  * Keep quality of original video on slipt.

 --  <pyarnau@gmail.com>  Sat, 17 Apr 2010 16:01:58 +0200

youtube-upload (0.2) unstable; urgency=low

  * Refactor video split
  * Create split file only when necessary
  * Add setup.py

 --  <pyarnau@gmail.com>  Tue, 16 Mar 2010 22:56:21 +0100

youtube-upload (0.1) unstable; urgency=high

  * Detect deprecated categories.
  * Validate meta-data before uploading the video.
  * Split files in chunks (youtube limits to 10') using ffmpeg.

 --  <pyarnau@gmail.com>  Tue, 08 Dec 2009 19:49:02 +0100

youtube-upload (0.0.1) unstable; urgency=high
 
  * First release.

 -- Arnau Sanchez <pyarnau@gmail.com>  Sat, 24 Jan 2009 12:17:21 +0200


youtube_upload/categories.py
import urllib2
import urllib
import json

URL = "https://www.googleapis.com/youtube/v3/videoCategories"

IDS = {
    "Film & Animation": 1,
    "Autos & Vehicles": 2,
    "Music": 10,
    "Pets & Animals": 15,
    "Sports": 17,
    "Short Movies": 18,
    "Travel & Events": 19,
    "Gaming": 20,
    "Videoblogging": 21,
    "People & Blogs": 22,
    "Comedy": 34,
    "Entertainment": 24,
    "News & Politics": 25,
    "Howto & Style": 26,
    "Education": 27,
    "Science & Technology": 28,
    "Nonprofits & Activism": 29,
    "Movies": 30,
    "Anime/Animation": 31,
    "Action/Adventure": 32,
    "Classics": 33,
    "Documentary": 35,
    "Drama": 36,
    "Family": 37,
    "Foreign": 38,
    "Horror": 39,
    "Sci-Fi/Fantasy": 40,
    "Thriller": 41,
    "Shorts": 42,
    "Shows": 43,
    "Trailers": 44,
}

def get(region_code="us", api_key=None):
    params = urllib.urlencode(dict(part="snippet", regionCode=region_code, key=api_key))
    full_url = URL + "?" + params
    response = urllib2.urlopen(full_url)
    categories_info = json.loads(response.read())
    items = categories_info["items"]
    categories = dict((item["snippet"]["title"], item["id"]) for item in items)
    return categories
